src/utils/mutate.js

```diff
@@ -0,0 +1,242 @@
+      // eslint-disable-next-line no-param-reassign
+      clone[key] = value;
+    }
+    return clone;
+  }, cloneDeep(operation));
+}
+
+// ----- write functions -----
+
+/**
+ *
+ * @param {object} firebase
+ * @param {Mutation_v1 | Mutation_v2} operation
+ * @param {Batch | Transaction} writer
+ * @returns {Promise | Doc} - Batch & Transaction .set returns null
+ */
+function write(firebase, operation = {}, writer = null) {
+  const { collection, path, doc, id, data, ...rest } = operation;
+  const ref = docRef(firebase.firestore(), path || collection, id || doc);
+  const changes = atomize(firebase, data || rest);
+
+  if (writer) {
+    const writeType = writer.commit ? 'Batching' : 'Transaction.set';
+    info(writeType, { id: ref.id, path: ref.parent.path, ...changes });
+    writer.set(ref, changes, { merge: true });
+    return { id: ref.id, path: ref.parent.path, ...changes };
+  }
+  info('Writing', { id: ref.id, path: ref.parent.path, ...changes });
+  return ref.set(changes, { merge: true });
+}
+
+/**
+ * @param {object} firebase
+ * @param {object} operations
+ * @returns {Promise}
+ */
+function writeSingle(firebase, operations) {
+  const done = mark('mutate.writeSingle');
+  const promise = write(firebase, operations);
+  done();
+  return promise;
+}
+
+const MAX_BATCH_COUNT = 500;
+
+/**
+ * @param {object} firebase
+ * @param {object} operations
+ * @returns {Promise}
+ */
+async function writeInBatch(firebase, operations) {
+  const done = mark('mutate.writeInBatch');
+  const committedBatchesPromised = chunk(operations, MAX_BATCH_COUNT).map(
+    (operationsChunk) => {
+      const batch = firebase.firestore().batch();
+      const writesBatched = operationsChunk.map((operation) =>
+        write(firebase, operation, batch),
+      );
+
+      return batch.commit().then(() => writesBatched);
+    },
+  );
+
+  done();
+  return Promise.all(committedBatchesPromised).then(flatten);
+}
+
+/**
+ * @param {object} firebase
+ * @param {object} operations
+ * @returns {Promise}
+ */
+async function writeInTransaction(firebase, operations) {
+  return firebase.firestore().runTransaction(async (transaction) => {
+    const serialize = (doc) =>
+      !doc
+        ? null
+        : { ...doc.data(), id: doc.ref.id, path: doc.ref.parent.path };
+    const getter = (ref) => {
+      info('Transaction.get ', { id: ref.id, path: ref.parent.path });
+      return transaction.get(ref);
+    };
+
+    const done = mark('mutate.writeInTransaction:reads');
+    const readsPromised = mapValues(operations.reads, (read) => {
+      if (isDocRead(read)) {
+        const doc = firestoreRef(firebase, read);
+        return getter(doc)
+          .then((snapshot) => (snapshot.exsits === false ? null : snapshot))
+          .then(serialize);
+      }
+
+      // NOTE: Firestore Transaction don't support collection inside
+      const coll = firestoreRef(firebase, read);
+      return coll
+        .get()
+        .then((snapshot) => {
```

Comment:
Since we are already in an async function we should be able to use await instead of mixing promise chaining and async/await

Reply:
Cool. Moved all .then into a async/await.